raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -592,6 +631,7 @@ private FetchQuorumRecordsResponseData handleFetchQuorumRecordsRequest(
                 .setNextFetchOffset(nextOffsetAndEpoch.offset)
                 .setNextFetchOffsetEpoch(nextOffsetAndEpoch.epoch);
         } else if (quorum.isVoter(replicaId)) {
+            maybeUpdateFetchTimerWithRemoteFetchTimestamp(state, replicaId, time.milliseconds());
```

Comment:
Seems like it would be more natural to pass both the fetch offset and timestamp into `updateReplicaEndOffset`. I'm guessing you didn't do this because of the returned value of `updateEndOffset`. I think a simple solution would be to expose a method to get the majority fetch timestamp and update blindly. Something like this:

```java
    private void updateReplicaEndOffset(LeaderState state, int replicaId, long endOffset, long timestamp) {
        if (state.updateEndOffset(replicaId, endOffset, timestamp)) {
            logger.trace("Leader high watermark updated to {} after replica {} end offset updated to {}",
                    state.highWatermark(), replicaId, endOffset);
            applyCommittedRecordsToStateMachine();
        }
        state.lastMajorityFetchTimestamp().ifPresent(timestamp -> 
           timer.resetDeadline(timestamp + fetchTimeoutMs);
        );
    }
```